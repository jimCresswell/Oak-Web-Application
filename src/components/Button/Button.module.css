/* button core */

.button {
  composes: button-unstyle from "../../styles/button-unstyle.module.css";
  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  font-weight: 600;
  max-width: 100%;
}

.button:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.button:not(:last-of-type) {
  margin-right: 12px;
}

/* button size */

.small {
  height: 40px;
  border-radius: 20px;
  padding: 0 12px;
}

.large {
  height: 48px;
  border-radius: 24px;
  padding: 0 16px;
}

/* button label */

.labelSpan {
  composes: ellipsis from "../../styles/ellipsis.module.css";
}

/* button icon */

.leadingIcon {
  flex-direction: row;
}

.trailingIcon {
  flex-direction: row-reverse;
}

.iconWrapper {
  display: inline-flex;
  align-items: center;
}

.leadingIcon .iconWrapper {
  margin-right: 8px;
}

.trailingIcon .iconWrapper {
  margin-left: 8px;
}

/* button variant */

.primary {
  background-color: var(--color-button-primary);
  color: var(--color-button-primary-contrast);
}

.secondary {
  background-color: var(--color-button-secondary);
  color: var(--color-button-secondary-contrast);
}

.tertiary {
  background-color: var(--color-button-tertiary);
  color: var(--color-button-tertiary-contrast);
}
