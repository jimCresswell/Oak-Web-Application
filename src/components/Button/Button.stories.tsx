import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import Component from "./Button";

export default {
  title: "Buttons/Button",
  component: Component,
  argTypes: {
    onClick: { action: "clicked" },
    label: {
      defaultValue: "Click me",
    },
    variant: {
      defaultValue: "brush",
    },
  },
} as ComponentMeta<typeof Component>;

const Template: ComponentStory<typeof Component> = (args) => (
  <Component {...args} />
);

export const Button = Template.bind({});

export const ButtonWithIcon = Template.bind({});
ButtonWithIcon.args = {
  icon: "Star",
};
