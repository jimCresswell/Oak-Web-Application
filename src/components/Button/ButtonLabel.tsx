import styled from "styled-components";

import ellipsis from "../../styles/ellipsis";

const ButtonLabel = styled.span`
  ${ellipsis}
`;

export default ButtonLabel;
