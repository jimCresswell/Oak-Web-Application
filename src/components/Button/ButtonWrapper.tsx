import styled, { css } from "styled-components";

import {
  ButtonSize,
  ButtonVariant,
  getButtonFlexDirection,
  getButtonHeight,
  getButtonPadding,
  getButtonBackground,
  getButtonColor,
  IconPosition,
} from "./common";

export type ButtonWrapperProps = {
  size: ButtonSize;
  iconPosition: IconPosition;
  variant: ButtonVariant;
};
const ButtonWrapper = styled.div<ButtonWrapperProps>`
  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  font-weight: 600;
  max-width: 100%;

  ${(props) => css`
    flex-direction: ${getButtonFlexDirection(props.iconPosition)};
    height: ${getButtonHeight(props.size)}px;
    border-radius: ${getButtonHeight(props.size) / 2}px;
    padding: 0 ${getButtonPadding(props.size)}px;
    background-color: ${getButtonBackground(props.variant)};
    color: ${getButtonColor(props.variant)};
  `}

  :disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
`;

export default ButtonWrapper;
