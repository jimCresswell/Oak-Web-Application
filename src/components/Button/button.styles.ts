import { css } from "styled-components";

import getColor from "../../styles/themeHelpers/getColor";
import margin, { MarginProps } from "../../styles/utils/spacing";

import {
  ButtonSize,
  ButtonVariant,
  getButtonFlexDirection,
  getButtonHeight,
  getButtonPadding,
  getButtonBackground,
  getButtonColor,
  IconPosition,
  CommonButtonProps,
  DEFAULT_BUTTON_VARIANT,
  DEFAULT_BUTTON_SIZE,
  DEFAULT_ICON_POSITION,
} from "./common";

export type ButtonStylesProps = MarginProps & {
  size: ButtonSize;
  iconPosition: IconPosition;
  variant: ButtonVariant;
};
export const getButtonStylesProps = (
  props: CommonButtonProps
): ButtonStylesProps => {
  const {
    variant = DEFAULT_BUTTON_VARIANT,
    iconPosition = DEFAULT_ICON_POSITION,
    size = DEFAULT_BUTTON_SIZE,
  } = props;

  return { size, iconPosition, variant };
};
const buttonStyles = css<ButtonStylesProps>`
  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  font-weight: 600;
  max-width: 100%;

  ${(props) => css`
    flex-direction: ${getButtonFlexDirection(props.iconPosition)};
    height: ${getButtonHeight(props.size)}px;
    border-radius: ${getButtonHeight(props.size) / 2}px;
    padding: 0 ${getButtonPadding(props.size)}px;
    background-color: ${getButtonBackground(props.variant)};
    color: ${getButtonColor(props.variant)};
  `}

  :disabled {
    background-color: ${getColor((theme) => theme.button.disabled.background)};
    cursor: not-allowed;
  }

  ${margin}
`;

export default buttonStyles;
