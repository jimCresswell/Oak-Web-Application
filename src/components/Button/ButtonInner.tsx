import { FC } from "react";

import Icon, { IconName } from "../Icon";

import ButtonIconWrapper from "./ButtonIconWrapper";
import ButtonLabel from "./ButtonLabel";
import ButtonWrapper from "./ButtonWrapper";
import {
  buttonIconSizeMap,
  ButtonSize,
  ButtonVariant,
  DEFAULT_BUTTON_SIZE,
  DEFAULT_BUTTON_VARIANT,
  DEFAULT_ICON_POSITION,
  IconPosition,
} from "./common";

export type ButtonInnerProps = {
  label: string;
  variant?: ButtonVariant;
  icon?: IconName;
  iconPosition?: IconPosition;
  size?: ButtonSize;
};
const ButtonInner: FC<ButtonInnerProps> = (props) => {
  const {
    variant = DEFAULT_BUTTON_VARIANT,
    icon,
    iconPosition = DEFAULT_ICON_POSITION,
    size = DEFAULT_BUTTON_SIZE,
    label,
  } = props;
  return (
    <ButtonWrapper size={size} iconPosition={iconPosition} variant={variant}>
      {icon && (
        <ButtonIconWrapper iconPosition={iconPosition}>
          <Icon name={icon} size={buttonIconSizeMap[size]} />
        </ButtonIconWrapper>
      )}
      <ButtonLabel>{label}</ButtonLabel>
    </ButtonWrapper>
  );
};

export default ButtonInner;
