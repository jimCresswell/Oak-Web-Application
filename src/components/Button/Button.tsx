import { forwardRef, MouseEventHandler } from "react";
import styled from "styled-components";

import useMediaQuery from "../../hooks/useMediaQuery";
import UnstyledButton from "../UnstyledButton";

import button, {
  ButtonStylesProps,
  getButtonStylesProps,
} from "./button.styles";
import ButtonInner from "./ButtonInner";
import {
  CommonButtonProps,
  defaultButtonProps,
  HTMLButtonProps,
} from "./common";
import IconButton from "./IconButton";

const StyledButton = styled(UnstyledButton)<ButtonStylesProps>`
  ${button}
`;

export type ButtonProps = CommonButtonProps & {
  onClick: MouseEventHandler<HTMLButtonElement>;
  htmlButtonProps?: HTMLButtonProps;
};

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    onClick,
    label,
    icon,
    "aria-label": ariaLabel,
    htmlButtonProps = {},
    iconBackground,
    iconOnlyOnMobile,
    ...spacingProps
  } = props;

  const isMobile = useMediaQuery("mobile");
  if (icon && iconOnlyOnMobile && isMobile) {
    return (
      <IconButton
        ref={ref}
        icon={icon}
        aria-label={ariaLabel || label}
        {...props}
        background={iconBackground}
      />
    );
  }

  const { size, variant, iconPosition, background } =
    getButtonStylesProps(props);

  return (
    <StyledButton
      ref={ref}
      {...htmlButtonProps}
      title={htmlButtonProps.title || ariaLabel || label}
      aria-label={ariaLabel || label}
      onClick={onClick}
      size={size}
      variant={variant}
      iconPosition={iconPosition}
      background={background}
      {...spacingProps}
    >
      <ButtonInner
        label={label}
        icon={icon}
        iconPosition={iconPosition}
        iconBackground={iconBackground}
        size={size}
        variant={variant}
        background={background}
      />
    </StyledButton>
  );
});

Button.defaultProps = defaultButtonProps;

export default Button;
