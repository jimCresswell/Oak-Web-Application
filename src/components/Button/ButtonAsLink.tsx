import { FC } from "react";
import Link, { LinkProps } from "next/link";
import styled from "styled-components";

import ButtonInner from "./ButtonInner";
import useButtonAsLinkProps from "./useButtonAsLinkProps";
import buttonStyles, {
  ButtonStylesProps,
  getButtonStylesProps,
} from "./button.styles";
import {
  CommonButtonProps,
  defaultButtonProps,
  HTMLAnchorProps,
} from "./common";

const StyledA = styled.a<ButtonStylesProps>`
  ${buttonStyles}
  ${({ disabled }) =>
    disabled === true &&
    `
   pointer-events: none;
  `}
`;
export type ButtonAsLinkProps = CommonButtonProps & {
  href: string;
  nextLinkProps?: Omit<LinkProps, "href">;
  htmlAnchorProps?: HTMLAnchorProps;
  disabled?: boolean;
};
const ButtonAsLink: FC<ButtonAsLinkProps> = (props) => {
  const {
    href,
    label,
    icon,
    "aria-label": ariaLabel,
    nextLinkProps,
    htmlAnchorProps = {},
    iconBackground,
    disabled = false,
    ...styleProps
  } = props;

  const { size, variant, iconPosition, background } =
    getButtonStylesProps(props);

  return (
    <Link {...nextLinkProps} href={href} passHref>
      <StyledA
        {...htmlAnchorProps}
        {...useButtonAsLinkProps()}
        title={htmlAnchorProps.title || ariaLabel || label}
        aria-label={ariaLabel || label}
        size={size}
        variant={variant}
        background={background}
        iconPosition={iconPosition}
        disabled={disabled}
        {...styleProps}
      >
        <ButtonInner
          label={label}
          icon={icon}
          iconPosition={iconPosition}
          iconBackground={iconBackground}
          size={size}
          background={background}
          variant={variant}
          disabled={disabled}
        />
      </StyledA>
    </Link>
  );
};

ButtonAsLink.defaultProps = defaultButtonProps;

export default ButtonAsLink;
