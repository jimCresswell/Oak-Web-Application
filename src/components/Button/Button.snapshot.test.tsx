import { render } from "@testing-library/react";

import Button from "./Button";

it("renders <Button> unchanged", () => {
  const { container } = render(
    <Button
      background="teachers-primary"
      href="/"
      label="Download"
      icon="Download"
    >
      Download
    </Button>
  );
  expect(container).toMatchSnapshot();
});
