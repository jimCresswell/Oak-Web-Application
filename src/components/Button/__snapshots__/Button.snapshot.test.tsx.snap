// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Button> unchanged 1`] = `
<div>
  <button
    class="button background--teachers-primary variant--rounded"
    href="/"
  >
    <span
      class="button-label"
    >
      Download
    </span>
    <span
      class="button-icon-wrapper"
    >
      <span
        class="outer-wrapper"
        style="height: 16px; width: 16px;"
      >
        <span
          class="inner-wrapper"
          style="height: 16px; width: 16px;"
        >
          <svg
            fill="none"
            height="100%"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2.5"
            viewBox="0 0 24 24"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
            />
            <polyline
              points="7 10 12 15 17 10"
            />
            <line
              x1="12"
              x2="12"
              y1="15"
              y2="3"
            />
          </svg>
        </span>
      </span>
    </span>
  </button>
</div>
`;
